distributed/nanny.py

```diff
@@ -245,15 +248,34 @@ def _():
         else:
             raise gen.Return('OK')
 
+    @gen.coroutine
     def memory_monitor(self):
         """ Track worker's memory.  Restart if it goes above 95% """
+        if self._memory_monitoring:
+            return
         if self.status != 'running':
             return
-        memory = psutil.Process(self.process.pid).memory_info().rss
-        frac = memory / self.memory_limit
-        if self.memory_terminate_fraction and frac > self.memory_terminate_fraction:
-            logger.warn("Worker exceeded 95% memory budget.  Restarting")
-            self.process.process.terminate()
+        self._memory_monitoring = True
+        try:
+            memory = psutil.Process(self.process.pid).memory_info().rss
+            frac = memory / self.memory_limit
+            if self.memory_terminate_fraction and frac > self.memory_terminate_fraction:
+                if hasattr(signal, 'SIGUSR1'):
```

Comment:
Might be nice to use a module-level constant for this, e.g. `GC_SIGNAL = getattr(signal, 'SIGUSR1', None)`